### PR TITLE
Issuer signing key validation: Remove exceptions

### PR DESCRIPTION
**Pull request by iNinja**
# Issuer signing key validation: Remove exceptions

- [x] You've read the [Contributor Guide](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/dev/Contributing.md) and [Code of Conduct](https://opensource.microsoft.com/codeofconduct/).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] If any gains or losses in performance are possible, you've included benchmarks for your changes. [More info](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/wiki/Benchmarking-in-Wilson)
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Removed exceptions as part of signing key validation.

## Description
- Implemented new (internal momentarily) method for signing key validation which replaces exception throwing with a return of the new SigningKeyValidationResult containing the exception information.
- Added new delegate signature for signing key validation
- Added tests around all the previous cases.

### CODE REVIEW

File: src/Microsoft.IdentityModel.Tokens/Validation/SigningKeyValidationResult.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        private Exception? _exception;
+
+        /// <summary>
+        /// Creates an instance of <see cref="SigningKeyValidationResult"/>
+        /// </summary>
+        /// <paramref name="signingKey"/> is the security key that was validated successfully.
+        public SigningKeyValidationResult(SecurityKey? signingKey)
+            : base(ValidationFailureType.ValidationSucceeded)
+        {
+            SigningKey = signingKey;
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Tokens/Validation/SigningKeyValidationResult.cs:23]: Can this actually be null in the success case?

File: src/Microsoft.IdentityModel.Tokens/Validation/Validators.IssuerSecurityKey.cs [hunk 2]
```diff
@@ -2,12 +2,37 @@
...
+#nullable enable
 namespace Microsoft.IdentityModel.Tokens
 {
+    /// <summary>
+    /// Definition for delegate that will validate the <see cref="SecurityKey"/> that signed a <see cref="SecurityToken"/>.
+    /// </summary>
+    /// <param name="signingKey">The security key to validate.</param>
+    /// <param name="securityToken">The <see cref="SecurityToken"/> that is being validated.</param>
+    /// <param name="validationParameters"><see cref="TokenValidationParameters"/> required for validation.</param>
+    /// <param name="callContext"></param>
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Tokens/Validation/Validators.IssuerSecurityKey.cs:21]: a few of these are empty

File: src/Microsoft.IdentityModel.Tokens/Validation/SigningKeyValidationResult.cs (hunk 1 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens/Validation/SigningKeyValidationResult.cs:23]: There is a case where the TVP have `ValidateIssuerSigningKey` set to `false` and we return in a successful manner while logging a warning about the fact that no validation has been performed. The security key can be `null` in this scenario without affecting the outcome.
This is a common pattern across our validators that we will need to decide if we want to differentiate as `NotEvaluated` or something similar. Right now the existing public methods return in the exact same manner if either the validation succeeds or no validation is performed (by not throwing an exception).

Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens/Validation/SigningKeyValidationResult.cs:23]: Also in the case where `TokenValidationParameters` have `RequireSignedTokens` set to `false`, a similar logic follows if the key is null.

File: src/Microsoft.IdentityModel.Tokens/Validation/Validators.IssuerSecurityKey.cs (hunk 2 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens/Validation/Validators.IssuerSecurityKey.cs:21]: Following up on our chat regarding all `CallContext` references.

File: src/Microsoft.IdentityModel.Tokens/Validation/SigningKeyValidationResult.cs (hunk 1 above)
Comment (keegan-caruso) [reply src/Microsoft.IdentityModel.Tokens/Validation/SigningKeyValidationResult.cs:23]: thanks

**Review by GeoK — APPROVED** [review]

**Review by keegan-caruso — APPROVED** [review]

**Review by FuPingFranco — APPROVED** [review]

**Pull request merged**